pilates/atlas/preprocessor.py

```diff
@@ -103,3 +106,115 @@ def prepare_atlas_inputs(settings, year, warm_start=False):
             except: 
                 logger.error('Urbansim Year {} Output Was Not Loaded Correctly by ATLAS'.format(year))
 
+
+
+logger = logging.getLogger(__name__)
+
+def compute_accessibility(path_list, measure_list, settings, year, threshold=500):
+    # set where to put atlas csv inputs (processed from urbansim outputs)
+    atlas_input_path = settings['atlas_host_input_folder'] + "/year{}".format(year)
+    
+    # for each OD, compute minimum time taken by public transit
+    # inf means no public transit available; unit = minute
+    ODmatrix = _get_time_ODmatrix(settings, path_list, measure_list, threshold)
+    
+    # assign values = 1 if time taken by public transit <= 30min; 0 if not
+    ODmatrix[ODmatrix<=30] = 1 
+    ODmatrix[ODmatrix>30] = 0
+    
+    # read and format geoid_to_zoneid mapping list
+    mapping = pd.read_csv('pilates/utils/data/{}/beam/geoid_to_zone.csv'.format(settings['region']))
+    mapping.index = mapping['GEOID']
+    mapping = mapping['zone_id'].to_dict()
+    
+    # read OD matrix size (i.e., range of zone_id)
+    skims_dir = settings['asim_local_input_folder']
+    skims = omx.open_file(os.path.join(skims_dir, 'skims.omx'), mode = 'r') 
+    zone_count = skims.shape()[0]
+    
+    # read in jobs data (keep low_memory=False to solve dtypeerror)
+    jobs = pd.read_csv("{}/year{}/jobs.csv".format(settings['atlas_host_input_folder'], year), low_memory=False)
+    
+    # map jobs geoid to zone id in OD matrix
+    jobs['zone_id'] = jobs['block_id'].map(mapping)
+    
+    # count number of jobs for each block_id
+    jobs_vector = jobs.groupby('block_id').agg({'job_id':'size','zone_id':'max'}).rename(columns={'job_id':'access_sum'})
+    
+    # average # of jobs per block for each taz
+    jobs_vector = jobs_vector.groupby('zone_id').agg({'access_sum':'mean'})
+    
+    # make sure every zone id has a row in jobs_vector
+    jobs_vector = jobs_vector.reindex(list(range(1,zone_count+1)), fill_value=0)
+    
+    # multiply OD matrix (o*d) with jobs vector (d*1)
+    # to get number of jobs accessible by public transit within 30min
+    accessibility = np.matmul(ODmatrix, jobs_vector)
+    accessibility.index.name = 'zone_id'
+    accessibility.index = accessibility.index + 1
+    
+    # # calculate taz-level zscore
+    # accessibility['access_zscore'] = (accessibility['access_sum'] - accessibility['access_sum'].mean())/accessibility['access_sum'].std()
+    
+    # # write taz-level accessibility data
+    # accessibility.to_csv('{}/accessibility_{}_taz.csv'.format(atlas_input_path, year))
+    
+    # read in taz_to_tract conversion matrix (1454*1588)
+    taz_to_tract = pd.read_csv('{}/taz_to_tract_{}.csv'.format(settings['atlas_host_input_folder'], settings['region']), index_col=0)
+    
+    # convert taz- to tract-level accessibility data
+    accessibility_tract = np.matmul(np.transpose(accessibility), np.array(taz_to_tract.values))
```

Comment:
Yes, one tract can be associated with multiple tazs, and one taz can be associated with multiple tracts. 

This tract-taz conversion has been carefully handled using the "taz_to_tract_sfbay.csv" file we generated externally. It is basically a weight matrix (dim = 1454 taz * 1588 tracts). Weight for taz n and tract m is the land area of taz n inside tract m, divided by the total land area in tract m. By doing np.matmul, a 1*1454 taz accessibility vector is multiplied by this matrix, and generates a 1588*1 tract accessibility vector. That said, the accessibility of a tract (i.e., number of jobs accessible from this tract) is estimated as the weighted average of accessibility of its overlapping tazs. 

It is noteworthy that this conversion relies on "taz_to_tract_sfbay.csv", so beamac (in settings.yaml) should only be set to 1 if running sfbay taz, otherwise we recommend setting beamac = 0. I can add an error message and comment in run.py file to clarify this issue.

Reply:
Ah, gotcha, thanks! I had assumed that tract-taz was a binary matrix but this makes a lot more sense and is a nice solution